src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp

```diff
@@ -540,24 +540,6 @@ MergeTreePrefetchedReadPool::ThreadsTasks MergeTreePrefetchedReadPool::createThr
     return result_threads_tasks;
 }
 
-MergeTreePrefetchedReadPool::~MergeTreePrefetchedReadPool()
-{
-    for (const auto & [_, thread_tasks] : threads_tasks)
-    {
-        for (const auto & task : thread_tasks)
-        {
-            if (task->reader.valid())
-                task->reader.wait();
```

Comment:
@kssenii I was looking into this too, so just to ensure that I'm we are on the same page, do I understand correctly that the problem was that not all features are kept in the `MergeTreePrefetchedReadPool` (`getTask` moves it out), hence some of the got lost and not waited properly?

Reply:
Yes, correct, according to the logs of stress test, the query was killed by the memory overcommit tracker and that is why it was not waited properly for some of the tasks which were already collected with getTask, but the same can also happen if some exception is thrown after the thread collected the task with getTask and before extracting the readers.